Adding libs for OpenXL

ishitaR88 opened the pull request.

zl-wang commented:
Two high-level comments:
1) provide a proper description of what you are trying to do in this PR;
2) you cannot modify the code as you did, since it breaks the builds using xlC compiler.  keep in mind that the codebase feeds into java8/11/17/21/newer builds. we only intended to switch to OpenXL for 'newer' builds like java23. if you change the code as you did in this PR, java8-21 cannot be built anymore (with xlC). you have to change the code (and makefile etc) conditionally to be compatible with both compilers.

zl-wang commented:
please indicate if it cleared the criteria that, with these changes in place, we can build with both xlC/xlclang and OpenXL (also staying or better the performance).  that is when it is ready to be reviewed.

zl-wang commented:
did I miss your changes to makefiles, to set CC and CXX?  at least, i haven't seen them yet.

pshipton commented:
I tried a build with the OMR (https://github.com/eclipse/omr/pull/7447) and these OpenJ9 changes.
https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK23_ppc64_aix_Personal/37
It got this far.
I added the compiler path (only works on paix822) and defined CC and CXX https://github.ibm.com/Peter-Shipton/tooling/tree/openxl
I had to fix (hack) openssl.gmk as well as revert the change to use xlc https://github.com/pshipton/openj9-openjdk-jdk23/tree/openxl
```
16:49:15  [ 41%] Running preprocessing p/runtime/ebb.spp to create /home/jenkins/workspace/Build_JDK23_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/runtime/compiler/p/runtime/ebb.ipp
16:49:15  error: invalid value 'extc99' in '-std=extc99'
```

pshipton commented:
> did I miss your changes to makefiles, to set CC and CXX?

defaults.yml should be updated similarly to https://github.ibm.com/Peter-Shipton/tooling/commit/eebd2986b8a3b

It won't work on any OpenJ9 open machine until we update them to have the 17.1.2 compiler.

ishitaR88 commented:
> I tried a build with the OMR ([eclipse/omr#7447](https://github.com/eclipse/omr/pull/7447)) and these OpenJ9 changes. https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK23_ppc64_aix_Personal/37 I added the compiler path (only works on paix822) and defined CC and CXX https://github.ibm.com/Peter-Shipton/tooling/tree/openxl I had to fix (hack) openssl.gmk as well as revert the change to use xlc https://github.com/pshipton/openj9-openjdk-jdk23/tree/openxl
> 
> It got this far:
> 
> ```
> 16:49:15  [ 41%] Running preprocessing p/runtime/ebb.spp to create /home/jenkins/workspace/Build_JDK23_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/runtime/compiler/p/runtime/ebb.ipp
> 16:49:15  error: invalid value 'extc99' in '-std=extc99'
> ```

Yes there was an issue. I have fixed that.will push the modification shortly.

pshipton commented on runtime/tests/redirector/jep178/module.xml:
```diff
@@ -34,7 +34,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 				<include-if condition="spec.zos_390.*"/>
 			</makefilestub>
-			<makefilestub data="UMA_CC_MODE += -qpic -brtl -bexpall">
+			<makefilestub data="UMA_CC_MODE += -fpic -brtl -bexpall">
```
This can't be changed unconditionally.
FYI module.xml (UMA) is mostly used for IBM Java 8 builds, but it's still (was) possible use UMA to build OpenJ9 versions, It's something we should check, if UMA is still working for Java 23 with OpenXL. By default we use cmake, but UMA can be used if adding options for debugging purposes.

keithc-ca commented on runtime/gc_glue_java/configure_includes/configure_aix_ppc.mk:
```diff
@@ -77,6 +77,10 @@ CONFIGURE_ARGS += 'OMR_HOST_OS=aix'
-
-CONFIGURE_ARGS+= 'GLOBAL_CFLAGS=-qstackprotect'
-CONFIGURE_ARGS+= 'GLOBAL_CXXFLAGS=-qstackprotect'
+ifeq ($(OMR_ENV_OPENXL),1)
+	CONFIGURE_ARGS+= 'GLOBAL_CFLAGS=-fstack-protector'
+	CONFIGURE_ARGS+= 'GLOBAL_CXXFLAGS=-fstack-protector'
+else
+	CONFIGURE_ARGS+= 'GLOBAL_CFLAGS=-qstackprotect'
+	CONFIGURE_ARGS+= 'GLOBAL_CXXFLAGS=-qstackprotect'
+endif
```
Lines in makefiles should not begin with a tab except within a script; please insert a space before each `+=` (see line 79):
```
ifeq ($(OMR_ENV_OPENXL),1)
  CONFIGURE_ARGS += 'GLOBAL_CFLAGS=-fstack-protector'
  CONFIGURE_ARGS += 'GLOBAL_CXXFLAGS=-fstack-protector'
else
  CONFIGURE_ARGS += 'GLOBAL_CFLAGS=-qstackprotect'
  CONFIGURE_ARGS += 'GLOBAL_CXXFLAGS=-qstackprotect'
endif
```

ishitaR88 replied:
I have added the space before +=. I initially used the tab as the previous ifeq condition checkings also used tabs instead of space. Instead of tab now I added 4 spaces

keithc-ca commented on runtime/j9vm/jvm.c:
```diff
@@ -4399,7 +4399,7 @@ JVM_Available(jint descriptor, jlong* bytes)
 	}
 
-#if defined(WIN32) && !defined(__IBMC__)
+#if defined(WIN32) && !defined(__IBMC__) && !(defined(__open_xl__) && !defined(__cplusplus))
```
In a C source file `!defined(__cplusplus)` is redundant.

zl-wang replied:
wondering why we need to do this to begin with. WIN32 definition already excludes the code from being picked up on AIX or zOS, what is the purpose of && with more other conditions?

keithc-ca commented on runtime/jilgen/CMakeLists.txt:
```diff
@@ -30,12 +30,22 @@ j9vm_add_executable(constgen
+			j9vm_interface
+			omrport
+			j9thr
+			j9util
```
If `j9util` is needed, it should not be conditional.

keithc-ca commented on runtime/makelib/targets.mk.aix.inc.ftl:
```diff
@@ -98,13 +102,20 @@ else
   CFLAGS += -qxlcompatmacros
 endif
-CFLAGS += -qlanglvl=extended -qarch=ppc -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108 -qstackprotect
+ifneq ($(OMR_ENV_OPENXL),1)
+  CFLAGS += -qlanglvl=extended -qarch=ppc -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108 -qstackprotect
+else
+  #openxl options
+  CFLAGS += -std=c89 -qarch=ppc -fno-strict-aliasing -fstack-protector
+endif
```
Please invert this to avoid the negation:
```
ifeq ($(OMR_ENV_OPENXL),1)
  # openxl options
  CFLAGS += -std=c89 -qarch=ppc -fno-strict-aliasing -fstack-protector
else
  CFLAGS += -qlanglvl=extended -qarch=ppc -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108 -qstackprotect
endif
```
Same comment near lines 128 & 139.

keithc-ca commented on runtime/makelib/targets.mk.aix.inc.ftl:
```diff
@@ -178,7 +203,9 @@ $(patsubst %.s,%.o,$(filter %.s,$(UMA_FILES_TO_PREPROCESS))) : %$(UMA_DOT_O) : %
   ifndef UMA_SUPPRESS_WARNINGS_AS_ERRORS
-    CFLAGS += -qhalt=w
-    CXXFLAGS += -qhalt=w
+    ifneq ($(OMR_ENV_OPENXL),1)
```
Please invert and add equivalent options for OpenXL.

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -1579,9 +1579,10 @@ typedef struct J9ExceptionHandler {
 } J9ExceptionHandler;
 
-#if defined(__xlC__) || defined(J9ZOS390)  /* Covers: Z/OS, AIX, Linux PPC*/
+//JACKIE: not sure where defined(__open_xl__) condition should go
```
C++ style comments cannot be used in a C header file (not that this one should stay anyway).

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -1579,9 +1579,10 @@ typedef struct J9ExceptionHandler {
+#if defined(__xlC__) || defined(J9ZOS390) || defined(__open_xl__)  /* Covers: Z/OS, AIX, Linux PPC*/
 #pragma pack(1)
-#elif defined(__ibmxl__) || defined(__GNUC__) || defined(_MSC_VER) /* Covers: Linux PPC LE, Windows, Linux x86 */
+#elif defined(__ibmxl__) || defined(__open_xl__) || defined(__GNUC__) || defined(_MSC_VER) /* Covers: Linux PPC LE, Windows, Linux x86 */
```
I think `defined(__open_xl__)` belongs on line 1583 which make this one dead.

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -1579,9 +1579,10 @@ typedef struct J9ExceptionHandler {
 
-#if defined(__xlC__) || defined(J9ZOS390)  /* Covers: Z/OS, AIX, Linux PPC*/
+//JACKIE: not sure where defined(__open_xl__) condition should go
+#if defined(__xlC__) || defined(J9ZOS390) || defined(__open_xl__)  /* Covers: Z/OS, AIX, Linux PPC*/
```
Please finish each sentence with a period, spell z/OS correctly and sort platforms:
```c
#if defined(__xlC__) || defined(J9ZOS390) || defined(__open_xl__) /* Covers: AIX, Linux PPC and z/OS. */
```

pshipton commented:
Tried a new build https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK23_ppc64_aix_Personal/52
There are a number of errors like
```
14:27:22  /home/jenkins/workspace/Build_JDK23_ppc64_aix_Personal/openj9/runtime/compiler/optimizer/InterpreterEmulator.cpp:1594:105: error: too many arguments to function call, expected 0, have 1
14:27:22   1594 |                                                       (int32_t) _currentCallMethod->virtualCallSelector(cpIndex), cpIndex, _currentCallMethod,
14:27:22        |                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^~~~~~~
```

ishitaR88 commented on runtime/cmake/platform/os/aix.cmake:
```diff
@@ -23,5 +23,10 @@
-set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-G,-brtl,-bernotok,-bnoentry,-bnolibpath")
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-brtl")
+if(CMAKE_C_COMPILER_IS_OPENXL)
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-G,-brtl,-bernotok,-bnoentry,-bnolibpath -latomic -lperfstat -liconv")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-brtl -latomic -lperfstat -liconv")
+else()
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-G,-brtl,-bernotok,-bnoentry,-bnolibpath")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-brtl")
```
I used tab for indentation

keithc-ca commented on runtime/compiler/env/alloca_openxl.h:
```diff
@@ -0,0 +1,3 @@
+#if defined(__open_xl__)
```
This file needs a copyright notice.

ishitaR88 replied:
added

ishitaR88 replied:
Changed year to 2024

zl-wang commented on runtime/compiler/runtime/asmprotos.h:
```diff
@@ -28,7 +28,7 @@ extern "C" {
 
 
-#if (defined(__IBMCPP__) || defined(__IBMC__) && !defined(MVS)) && !defined(J9ZOS390) && !defined(LINUXPPC64)
+#if (defined(__IBMCPP__) || defined(__IBMC__) && !defined(MVS)) && !defined(J9ZOS390) && !defined(LINUXPPC64) || defined(__open_xl__)
```
this looks weird, given the fact that && has higher precedence over ||.  it seems dangerous or unreadable to mix up compiler macro (__open_xl__) with platform macro (J9ZOS390 and LINUXPPC64).

zl-wang replied:
@r30shah this looks an existing problem as well re MVS.  i think MVS is also a platform macro (i.e. not defined by the compiler. but defined to indicate the particular platform). and, I think it should be `__MVS__` to begin with.  if @r30shah agrees, we should fix this as well.

r30shah replied:
I agree, it should be `__MVS__` and with the fix Julian suggests.

ishitaR88 replied:
As the MVS issue is not introduced by the changes of this PR I will address this in future creating a new PR.

zl-wang commented on runtime/libffi/preconf/xz64/ffi.h:
```diff
@@ -96,7 +96,7 @@ extern "C" {
 #  ifdef _AIX /* AIX 5.1 and earlier have LONGLONG_MAX */
 #   ifndef __PPC64__
-#    if defined (__IBMC__) || defined (__IBMCPP__)
+#    if defined (__IBMC__) || defined (__IBMCPP__) || defined(__open_xl__)
```
for the above dozen ffi.h files on non-AIX platforms, do we really need to change this line of code, especially since it is under` #  ifdef _AIX ` to begin with.  in other words, how changing or not-changing this code affect the builds on non-AIX platforms?

zl-wang commented on runtime/makelib/targets.mk.aix.inc.ftl:
```diff
@@ -98,13 +102,22 @@ else
+else
+  CFLAGS += -qlanglvl=extended -qarch=ppc -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108 -qstackprotect
+endif
 CFLAGS += -D_XOPEN_SOURCE_EXTENDED=1 -D_ALL_SOURCE -DRS6000 -DAIXPPC -D_LARGE_FILES
```
take notice maybe:  we need to revisit this flag setting in the future ... -qarch=ppc is pretty out of date and conservative.

ishitaR88 commented on runtime/compiler/runtime/RelocationRecord.cpp:
```diff
@@ -64,7 +64,7 @@
 
 // TODO: move this someplace common for RuntimeAssumptions.cpp and here
-#if defined(__IBMCPP__) && !defined(AIXPPC) && !defined(LINUXPPC)
+#if (defined(__IBMCPP__) || defined(__open_xl__) && defined(__cplusplus)) && !defined(AIXPPC) && !defined(LINUXPPC)
```
Fixed it in new commit

ishitaR88 commented on runtime/j9vm/jvm.c:
```diff
@@ -4542,7 +4542,7 @@ JVM_Lseek(jint descriptor, jlong bytesToSeek, jint origin)
 
 #if defined(WIN32)
-#ifdef __IBMC__
+#if defined(__IBMC__) || defined(__open_xl__) && !defined(__cplusplus)
```
changed

keithc-ca commented on runtime/j9vm/jvm.c:
```diff
@@ -4642,7 +4642,7 @@ JVM_Open(const char* filename, jint flags, jint mode)
 
 #ifdef WIN32
-#ifdef __IBMC__
+#if defined(__IBMC__) || defined(__open_xl__) && !defined(__cplusplus)
```
Please remove redundant uses of `defined(__cplusplus)` throughout.

keithc-ca commented on runtime/libffi/preconf/arm/ffi.h:
```diff
@@ -96,7 +96,7 @@ extern "C" {
 #  ifdef _AIX /* AIX 5.1 and earlier have LONGLONG_MAX */
 #   ifndef __PPC64__
-#    if defined (__IBMC__) || defined (__IBMCPP__)
+#    if defined (__IBMC__) || defined (__IBMCPP__) || defined(__open_xl__)
```
Please revert changes such as this (where Open XL can't apply).

keithc-ca commented on runtime/makelib/targets.mk.aix.inc.ftl:
```diff
@@ -114,13 +127,26 @@ else
-  CXXFLAGS += -q mbcs -qinfo=pro
+  ifeq ($(OMR_ENV_OPENXL),1)
+    # for openxl do nothing
+  else
+    # xlc++ options
+    CXXFLAGS += -q mbcs -qinfo=pro
+  else
```
This is not legal (only one `else` is allowed).

keithc-ca commented on runtime/verbose/CMakeLists.txt:
```diff
@@ -62,25 +62,52 @@ target_include_directories(j9vrb
-		j9zip
-		j9utilcore
-)
+if(DEFINED OMR_OS_AIX AND DEFINED CMAKE_C_COMPILER_IS_OPENXL)
+	target_link_libraries(j9vrb
+		PRIVATE
```
Please don't do this. If libraries are need by any compiler, they should be used for all.

keithc-ca commented:
The description is incomplete. This isn't just about libraries - many command-line options are adjusted and source is changed.

Please follow the commit guidelines at https://github.com/eclipse-omr/omr/blob/master/CONTRIBUTING.md#commit-guidelines (the commit summary and the headline here should be "written in the imperative mood").

midronij commented:
Ishita will be away for the foreseeable future, so I will be taking over these changes on this PR: https://github.com/eclipse-openj9/openj9/pull/20690
